cpp/src/community/edge_triangle_count_impl.cuh

```diff
@@ -78,77 +78,298 @@ struct update_edges_p_r_q_r_num_triangles {
-      edge_first});
-
-  return num_triangles;
+  // auto my_rank = handle.get_comms().get_rank();
+  if constexpr (multi_gpu) {
+    num_chunks = host_scalar_allreduce(
+      handle.get_comms(), num_chunks, raft::comms::op_t::MAX, handle.get_stream());
+  }
+
+  // Need to ensure that the vector has its values initialized to 0 before incrementing
+  thrust::fill(handle.get_thrust_policy(), num_triangles.begin(), num_triangles.end(), 0);
+
+  for (size_t i = 0; i < num_chunks; ++i) {
+    auto chunk_size = std::min(edges_to_intersect_per_iteration, num_remaining_edges);
+    num_remaining_edges -= chunk_size;
+    // Perform 'nbr_intersection' in chunks to reduce peak memory.
+    auto [intersection_offsets, intersection_indices] =
+      detail::nbr_intersection(handle,
+                               graph_view,
+                               cugraph::edge_dummy_property_t{}.view(),
+                               edge_first + prev_chunk_size,
+                               edge_first + prev_chunk_size + chunk_size,
+                               std::array<bool, 2>{true, true},
+                               false /*FIXME: pass 'do_expensive_check' as argument*/);
+
+    // Update the number of triangles of each (p, q) edges by looking at their intersection
+    // size
+    thrust::for_each(
+      handle.get_thrust_policy(),
+      thrust::make_counting_iterator<edge_t>(0),
+      thrust::make_counting_iterator<edge_t>(chunk_size),
+      [chunk_start          = prev_chunk_size,
+       num_triangles        = raft::device_span<edge_t>(num_triangles.data(), num_triangles.size()),
+       intersection_offsets = raft::device_span<size_t const>(
+         intersection_offsets.data(), intersection_offsets.size())] __device__(auto i) {
+        num_triangles[chunk_start + i] += (intersection_offsets[i + 1] - intersection_offsets[i]);
+      });
+
+    if constexpr (multi_gpu) {
+      // stores all the pairs (p, r) and (q, r)
+      auto vertex_pair_buffer_tmp = allocate_dataframe_buffer<thrust::tuple<vertex_t, vertex_t>>(
+        intersection_indices.size() * 2, handle.get_stream());  // *2 for both (p, r) and (q, r)
+      // So that you shuffle only once
+
+      // tabulate with the size of intersection_indices, and call binary search on
+      // intersection_offsets to get (p, r).
+      thrust::tabulate(
+        handle.get_thrust_policy(),
+        get_dataframe_buffer_begin(vertex_pair_buffer_tmp),
+        get_dataframe_buffer_begin(vertex_pair_buffer_tmp) + intersection_indices.size(),
+        extract_p_r_q_r<vertex_t, edge_t, decltype(edge_first)>{
+          prev_chunk_size,
+          0,
+          raft::device_span<size_t const>(intersection_offsets.data(), intersection_offsets.size()),
+          raft::device_span<vertex_t const>(intersection_indices.data(),
+                                            intersection_indices.size()),
+          edge_first});
+      // FIXME: Consolidate both functions
```

Comment:
And please address FIXMEs you can quickly address. We can leave performance optimizations or something that requires new primitives for future FIXMEs, but something you can independently and quickly address, we should address.

Reply:
Right. And this FIXME was already addressed. Prior to this, I was launching two separate kernels to extract `p, r` and `q, r` edges which I consolidated to `extract_p_r_q_r`. I will remove the FIXME in my next commit